Custom dataloader registry support

ori-kron-wis opened the pull request.

canergen commented on .github/workflows/test_linux.yml:
```diff
@@ -53,7 +53,7 @@ jobs:
           COLUMNS: 120
         run: |
-          coverage run -m pytest -v --color=yes
+          coverage run -m pytest -v --color=yes -m "not custom_dataloader"
```
what's this?

canergen commented on src/scvi/model/_scanvi.py:
```diff
@@ -128,23 +129,29 @@ def __init__(
+                if REGISTRY_KEYS.CAT_COVS_KEY in self.adata_manager.data_registry
+                else None
+            )
+        else:
```
We should include the case that CAT_COVS_KEY is in registry.

canergen commented on src/scvi/model/_scanvi.py:
```diff
@@ -128,23 +129,29 @@ def __init__(
+        use_size_factor_key = self.registry_["setup_args"][f"{REGISTRY_KEYS.SIZE_FACTOR_KEY}_key"]
         library_log_means, library_log_vars = None, None
-        if not use_size_factor_key and self.minified_data_type is None:
+        if self.adata is not None and not use_size_factor_key and self.minified_data_type is None:
```
Can you create a seperate PR and merge it that changes the if statement to use_observed_lib_size=False (see vae code).

canergen commented on src/scvi/model/_scanvi.py:
```diff
@@ -250,7 +260,8 @@ def from_scvi_model(
         )
-        scanvi_model = cls(adata, **non_kwargs, **kwargs, **scanvi_kwargs)
+
+        scanvi_model = cls(adata, scvi_model.registry, **non_kwargs, **kwargs, **scanvi_kwargs)
```
Is it supported to have both adata and registry?

canergen commented on src/scvi/model/_scvi.py:
```diff
@@ -147,46 +148,42 @@ def __init__(
-                **kwargs,
-            )
-            self.module.minified_data_type = self.minified_data_type
+        else:
```
I don't think we need the if statement.

canergen commented on src/scvi/model/_scvi.py:
```diff
@@ -147,46 +148,42 @@ def __init__(
+        n_batch = self.summary_stats.n_batch
+        use_size_factor_key = self.get_setup_arg(f"{REGISTRY_KEYS.SIZE_FACTOR_KEY}_key")
+        library_log_means, library_log_vars = None, None
+        if self.adata is not None and not use_size_factor_key and self.minified_data_type is None:
```
See above, change this check. For customdataloader we should assert that it's not computing library_log_means and vars this is not supported. Currently it would give an error if use_observed_lib_size=False

canergen commented on src/scvi/model/_scvi.py:
```diff
@@ -232,6 +229,112 @@ def setup_anndata(
         cls.register_manager(adata_manager)
 
+    @staticmethod
+    def _get_summary_stats_from_registry(registry: dict) -> attrdict:
```
This should be in base or somewhere else. It doesn't make sense in scVI.

canergen commented on src/scvi/model/_scvi.py:
```diff
@@ -232,6 +229,112 @@ def setup_anndata(
+
+    @classmethod
+    @setup_anndata_dsp.dedent
+    def setup_datamodule(
```
This needs to be defined for each model. Can we reduce the amount of duplicate code by adding something similar to setup_anndata?

ori-kron-wis replied:
I guess you mean using data fields. Let's remember this once everything works.

canergen commented on src/scvi/model/_scvi.py:
```diff
@@ -232,6 +229,112 @@ def setup_anndata(
+    @setup_anndata_dsp.dedent
+    def setup_datamodule(
+        cls,
+        datamodule,  # TODO: what to put here?
```
It should be pytorch.DataLoader right? Martin has done typing for it in the current code.

canergen commented on src/scvi/model/_scvi.py:
```diff
@@ -232,6 +229,112 @@ def setup_anndata(
+                    "state_registry": {
+                        "n_obs": datamodule.n_obs,
+                        "n_vars": datamodule.n_vars,
+                        "column_names": [str(i) for i in column_names],  # TODO: from adata (czi)?
```
Not following?

canergen commented on src/scvi/model/base/_archesmixin.py:
```diff
@@ -82,56 +87,64 @@ def load_query_data(
-                _validate_var_names(adata[modality], var_names[modality])
+                    logger.debug("Subsetting query vars to reference vars.")
+                    adata._inplace_subset_var(var_names)
+                _validate_var_names(adata, var_names)
```
We need to verify for dataloaders that the gene names are matching.

canergen replied:
And the order.

canergen commented on src/scvi/model/base/_archesmixin.py:
```diff
@@ -82,56 +87,64 @@ def load_query_data(
-            logger.debug("Subsetting query vars to reference vars.")
-            adata._inplace_subset_var(var_names)
-        _validate_var_names(adata, var_names)
+            registry = attr_dict.pop("registry_")
```
This check and the ones below are independent of datamodule or AnnData, right? Remove the indent.

canergen replied:
I am not sure why every code here is displayed as modified?

ori-kron-wis replied:
beucase of "if adata:" in line 108 all is indented.

canergen commented on src/scvi/model/base/_archesmixin.py:
```diff
@@ -202,7 +215,7 @@ def prepare_query_anndata(
         in which case a pd.Index of reference var names is returned.
         """
-        _, var_names, _ = _get_loaded_data(reference_model, device="cpu")
+        _, var_names, _ = _get_loaded_data(reference_model, device="cpu", adata=adata)
```
Does this work with a dataloader?

canergen commented on src/scvi/model/base/_archesmixin.py:
```diff
@@ -350,15 +363,15 @@ def requires_grad(key):
 
 
-def _get_loaded_data(reference_model, device=None):
+def _get_loaded_data(reference_model, device=None, adata=None):
```
Why do we need adata here?

canergen commented on src/scvi/model/base/_base_model.py:
```diff
@@ -105,10 +121,20 @@ def __init__(self, adata: AnnOrMuData | None = None):
+            self.registry_ = registry
+            self.summary_stats = _get_summary_stats_from_registry(registry)
+        elif self.__class__.__name__ == "GIMVI":
+            # note some models do accept empty registry/adata (e.g: gimvi)
```
I'm not following this one. What is the exception with GIMVI?

canergen commented on src/scvi/model/base/_base_model.py:
```diff
@@ -242,6 +282,40 @@ def _register_manager_for_instance(self, adata_manager: AnnDataManager):
+        else:
+            return self._adata_manager.get_from_registry(registry_key)
+
+    # def get_from_registry(self, registry_key: str) -> np.ndarray | pd.DataFrame:
```
what's this?

canergen commented on src/scvi/model/base/_base_model.py:
```diff
@@ -465,13 +542,20 @@ def _validate_anndata(
 
+    def transfer_fields(self, adata: AnnOrMuData, **kwargs) -> AnnData:
+        """Transfer fields from a model to an AnnData object."""
+        if self.adata:
```
where do we need transfer_fields? can we make it work with datamodule?

canergen commented on src/scvi/model/base/_base_model.py:
```diff
@@ -627,8 +711,7 @@ def save(
         model_state_dict = self.module.state_dict()
-
-        var_names = _get_var_names(self.adata, legacy_mudata_format=legacy_mudata_format)
+        var_names = self.get_var_names(legacy_mudata_format=legacy_mudata_format)
```
do we need two get_var_names function?

canergen commented on src/scvi/model/base/_base_model.py:
```diff
@@ -705,31 +789,32 @@ def load(
+                    "Saved model does not contain original setup inputs. "
+                    "Cannot load the original setup."
+                )
+            _validate_var_names(adata, var_names)
```
should be validated also for a dataloader.

canergen replied:
Remove if adata statement in 791.

canergen commented on src/scvi/model/base/_base_model.py:
```diff
@@ -885,18 +970,164 @@ def view_anndata_setup(
+
+    def get_state_registry(self, registry_key: str) -> attrdict:
+        """Returns the state registry for the AnnDataField registered with this instance."""
+        return attrdict(self.registry_[_FIELD_REGISTRIES_KEY][registry_key][_STATE_REGISTRY_KEY])
```
Does this work with dataloader. Documentation should be updated then.

canergen commented on src/scvi/model/base/_save_load.py:
```diff
@@ -133,7 +133,10 @@ def _initialize_model(cls, adata, attr_dict):
 
-    model = cls(adata, **non_kwargs, **kwargs)
+    if not adata:
+        adata = None
```
Is adata false here? Do we need a default value for registry?

canergen commented on src/scvi/model/base/_training_mixin.py:
```diff
@@ -84,27 +84,14 @@ def train(
-
         if max_epochs is None:
-            if datamodule is None:
+            if self.adata is not None:
```
We should take here n_obs from summary stats to make it compatible with a dataloader.

canergen replied:
See below we don't need the if statement.

canergen replied:
Should be fixed.

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+    experiment_name = "mus_musculus"
+    obs_value_filter = 'is_primary_data == True and tissue_general in ["kidney"] and nnz >= 3000'
+
+    # This is under comments just to save time (selecting highly varkable genes):
```
Remove this block, we don't need it.

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+        dataloader_kwargs={"num_workers": 0, "persistent_workers": False},
+    )
+
+    # table of genes should be filtered by soma_joinid - but we should keep the encoded indexes
```
no need to.

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+    _ = model.get_reconstruction_error(dataloader=dataloader)
+    _ = model.get_latent_representation(dataloader=dataloader)
+
+    scvi.model.SCVI.prepare_query_anndata(adata_orig, reference_model=model)
```
test this with a second model trained using dataloader

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+    n_layers = 1
+    n_latent = 50
+
+    scvi.model._scvi.SCVI.setup_datamodule(datamodule)  # takes time
```
I like this part. It's nice.

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+
+    pprint(datamodule.registry)
+
+    batch_size = 1024
```
does batch size have an effect. I thought it's defined by the datamodule?

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+    # _ = model_census2.get_latent_representation()
+
+    # takes time
+    adata = cellxgene_census.get_anndata(
```
just download 10 cells - see below with obs_value_filter.

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+        var_coords=hv_idx,
+    )
+
+    # TODO: do we need to put inside (or is it alrady pre-made) - perhaps need to tell CZI
```
I assume we need to make it.

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+    adata.obs["batch"] = adata.obs[batch_keys].agg("".join, axis=1).astype("category")
+
+    scvi.model.SCVI.prepare_query_anndata(adata, save_path)
+    scvi.model.SCVI.load_query_data(registry=datamodule.registry, reference_model=save_path)
```
We should have more tests that actually fail - using different genes without prepare_query_anndata and different batch categories. Assert that it fails.

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+
+    scvi.model.SCVI.prepare_query_anndata(adata, model_census2)
+
+    scvi.model.SCVI.setup_anndata(adata, batch_key="batch")  # needed?
```
checking that an AnnData model can be trained using datamodule. Do we really want it?

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+
+    user_attributes_model_census3 = model_census3._get_user_attributes()
+    pprint(user_attributes_model_census3)
+    _ = model_census3.get_elbo()
```
uses AnnData for inference?

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+    scvi.model.SCVI.prepare_query_anndata(adata, model_census3)
+    scvi.model.SCVI.load_query_data(adata, model_census3)
+
+    datamodule_inference = CensusSCVIDataModule(
```
check here that using different genes and different batches fails. You can take much fewer cells here, like 1000.

canergen commented on tests/dataloaders/test_custom_dataloader.py:
```diff
@@ -0,0 +1,400 @@
+    # Create a dataloder of a CZI module
+    datapipe = datamodule_inference.datapipe
+    dataloader = experiment_dataloader(datapipe, num_workers=0, persistent_workers=False)
+    mapped_dataloader = (
```
What's this?

canergen commented on tests/model/test_scvi.py:
```diff
@@ -1256,6 +1210,23 @@ def test_scvi_inference_custom_dataloader(n_latent: int = 5):
+
+    model = SCVI(adata, n_latent=n_latent)
+    model.train(max_epochs=1)
+    dataloader = model._make_data_loader(adata)
```
Does model._make_data_loader exist for all models? We should then add the test to the other models as well?

canergen replied:
Is the dataloader sufficient to also setup the model and does setup_datamodule work for it?

canergen commented on src/scvi/dataloaders/_custom_dataloders.py:
```diff
@@ -0,0 +1,194 @@
+import scvi
+
+if TYPE_CHECKING:
+    import lamindb as ln
```
does this create a requirement for lamindb? Can we make MappedCollectionDataModule a lazy load?

canergen commented on src/scvi/dataloaders/_custom_dataloders.py:
```diff
@@ -0,0 +1,194 @@
+                    "state_registry": {
+                        "categorical_mapping": self.label_keys,
+                        "original_key": self._label_key,
+                        "unlabeled_category": "unlabeled",
```
expose "unlabeled"?

canergen commented on src/scvi/model/_scanvi.py:
```diff
@@ -406,17 +442,24 @@ def train(
+                **datasplitter_kwargs,
+            )
+        else:
+            # TODO fix in external dataloader?
```
print warning.

canergen commented on src/scvi/model/_scvi.py:
```diff
@@ -134,6 +148,7 @@ def __init__(
         )
+        self._module_init_on_train = False
 
         if self._module_init_on_train:
```
remove module_init_on_train. This is deprecated code with the new dataloader implementation.

canergen commented on src/scvi/model/_scvi.py:
```diff
@@ -144,17 +159,29 @@ def __init__(
+                )
+            else:
+                # custom datamodule
+                n_cats_per_cov = self.summary_stats[f"n_{REGISTRY_KEYS.CAT_COVS_KEY}"]
```
does this work for AnnData? Can we remove the if statement here?

marianogabitto commented:
Hi Ori,
      I gave it a try to this branch. First, I needed to install psutil. Second, when I ran the tutorial, I found an error:

Code:

model = scvi.model.SCVI(adata=None, registry=datamodule.registry, datamodule=datamodule)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/allen/programs/celltypes/workgroups/rnaseqanalysis/Mariano/Anaconda3/envs/scvi-env-largedata/lib/python3.12/site-packages/scvi/model/_scvi.py", line 184, in __init__
    self.module = self._module_cls(
                  ^^^^^^^^^^^^^^^^^
TypeError: VAE.__init__() got an unexpected keyword argument 'datamodule'

marianogabitto commented:
I fixed it.

In the tutorial file: scvi-tools/docs/user_guide/use_case /custom_dataloaders.md

Replace the line:
model = scvi.model.SCVI(adata=None, registry=datamodule.registry, datamodule=datamodule)

with:
model = scvi.model.SCVI(adata=None, registry=datamodule.registry)

ori-kron-wis commented:
@marianogabitto thanks. I fixed those things. 
Also see the test_custom_dataloder.py file and the corresponding tutorial on this to see the full capabilities we have currently (https://github.com/scverse/scvi-tutorials/pull/425).

As you saw, my concern right now is the slowness of the process. Let's focus on this.
Did you find/ have any suggestions on how we could make it faster?

feel free to fork the branch and add your commits.

marianogabitto commented:
Hi Ori,
        I am adding the point that I brought in discourse.

        "I compared head to head the TileDB and the regular anndata loaders and Tile is 50% slower. I tested the TileDB dataloader ir regular and DDP mode and what is causing the delay is the slow access to data. GPU peaks and process super fast but in between batches there is a long waiting time "

        Let me tell you how I did this comparison. On the one hand, I save the TileDB experiment as an anndata and I run scvi regularly. On the other hand, I grab the data loader code from scvi (splitting, Anndataloader, AnnDataset) and create an scvi external AnnData loader (just to be sure there was no difference between running the anndata in regular scvi versus passing it as a new anndataloader).

2 questions:

        1) If you can wait until Wednesday, I am going to talk to people in my High Performance Computer Cluster and representatives from TileDB and try to debug if there is any way to accelerate the data loader with multiple workers or options. It is expected that the data loader that access disk is slower, I want to see how slow.

        2) what is the data loader that is used in regular scvi ? Should I also use the following? wouldn't the train dataloder only have an amount of data equivalent to train_size ?
        inference_dataloader = (
                  inference_datamodule.on_before_batch_transfer(batch, None)
                  for batch in inference_datamodule.train_dataloader()
        )

M

canergen commented:
Just checking is it still faster if you load the AmnData in disk backed mode? This would indeed be surprising while the other overhead could come from loading from disk? Any chance to use a fast SSD storage of the data as usually recommended for from-disk loading. Data could also be scattered across SSDs and there tileDB might have suggestions how to optimize this - not perfect randomness is not really an issue if they first randomize order on disks (not a single experiment on one SSD).

marianogabitto commented:
Speeds are:
anndata in memory > anndata backed >~ tileDB .

yes, I am testing two things: I am talking with my support team about a partition with fast disk access and touching base with tiled about how to optimize this.

ori-kron-wis commented:
@marianogabitto I will try to do some testing on my end.
Actually, the whole part of "on_before_batch_transfer" was something that I inherited from he census implementation long ago, but I never had the time to check it fully (see https://github.com/chanzuckerberg/cellxgene-census/blob/756708e9aa18791b7bae3712e9dd66d2b6ce9d75/api/python/notebooks/experimental/pytorch_scvi.ipynb)
and it makes sense to me, it might be the bottleneck.
We are not in a rush to release it.

marianogabitto commented:
But you need the "on_before_batch" to move the batch to PyTorch, no ? isn't the reason of that callback to get the X, batch, labels data and to move it to tensors? what happens if they are already tensors?

marianogabitto commented:
One more... how about making the anndata_manager and the registry tutorial ?

ori-kron-wis commented:
We will deal with everything, but let's take it step by step. on_before_batch is needed of course, but lets see how we can improve it.

ori-kron-wis commented:
Your suggestion worked and I do see much better performance in train time when using 
```
datamodule.setup()
model = scvi.model.SCVI(
    adata=None,
    registry=datamodule.registry,
    n_layers=n_layers,
    n_latent=n_latent,
    gene_likelihood="nb",
    encode_covariates=False,
)
# creating the dataloader for trainset
training_dataloader = (
    datamodule.on_before_batch_transfer(batch, None)
    for batch in datamodule.train_dataloader()
)
import time
start = time.time()
model.train(
    datamodule=training_dataloader,
    #datamodule=datamodule,
    max_epochs=100,
    batch_size=1024,
    # accelerator="gpu",
    # devices=-1,
    # strategy="ddp_find_unused_parameters_true",
)
end = time.time()
print(f"Elapsed time: {end - start:.2f} seconds")
```
See my updated tutorial on the other branch

marianogabitto commented:
Ori, I got lost ... what branch should I look for the tutorial? What branch should I check for testing your commits?

ori-kron-wis commented:
This are the tutorials:
https://app.reviewnb.com/scverse/scvi-tutorials/pull/425/

tests are in the current branch

marianogabitto commented:
Ori, this is not working for me. When I invoke in the notebook:
training_dataloader = (
    datamodule.on_before_batch_transfer(batch, None) for batch in datamodule.train_dataloader()
)
I get:
switching torch multiprocessing start method from "fork" to "spawn"
and then errors out....

marianogabitto commented:
Ori, can you monitor your GPU usage? This is how I am measuring speedup. It is not using GPU so much compare to data loader in memory.

ori-kron-wis commented:
Hi @marianogabitto ,
Thanks

1. I made several changes, and I added the on_before_batch_transform into the class, it is not part of the analysis code now. So if you pulled the branch and reinstalled, you will get errors for running the same code as before.
I have updated the tutorial, sorry for this.

But im not following on your code, can you share what you are running exactly so we can compare?

2. Regarding the GPU behavior. I see it the same.
The data you use matters (the speed enhancement is seen in larger data, not smaller ones).

I dont think GPU is not utilized, it just that the data load is much slower in tiledb, as you said 100 times slower in that sense. so while with adata the data loading is 1s we see almost continuous use of the GPU and in the tileDB s3 there's a 100sec gap between the same GPU usage, so we mistakenly see it underuse.

3. Num workers is for multiprocessing loading and is a parameter in the torch dataloader. We know that it is also dependent on data size and that we do not always get what we expect from it, specifically, there is overhead with initializing and closing it. 
How do you use it? I will try to check its speed in the custom dataloder context also, in any case, we should run it with the number that best benefits us, it's not a magic thing that helps each time.

4.  I think the common practice is a 1 GPU running on a notebook. We need to make sure this is working, and other scenarios will follow. 
But having said that, the scripts of running with DDP compared to running them in notebooks can be very different, and we need to test all possibilities. We might find that we need to run it as a script like this example. Will check.

5. I added SCANVI to the tutorials as well. Im a bit still struggling with its predictions, but integration is fine.

marianogabitto commented:
Ori,
     I am testing updates in 12 hours. Sorry for the delay. 
     One more thing in the meantime. It will be great to expose the scvi Anndata DataLoader as an example of what is going on internally. This code does not work because the BatchDistributedSampler is not outputting the samples with the correct dimensions (In DDP), but if you help me solve it, it will be great.

## Code ##
from scvi.dataloaders import DataSplitter

scvi.model.SCVI.setup_anndata(adata, batch_key="batch", categorical_covariate_keys=['cell_type', 'donor'])
ad_manager = scvi.model.SCVI._get_most_recent_anndata_manager(adata, required=True)

model = scvi.model.SCVI(
            registry=ad_manager._registry,
            gene_likelihood="nb",
            encode_covariates=False,
        )

ad_manager.adata = adata
dl = DataSplitter(ad_manager, train_size=0.9, pin_memory=True, num_workers=2, persistent_workers=True)#, prefetch_factor=2)
dl.setup()

model.train(
        datamodule=datamodule,
        max_epochs=10,
        batch_size=128,
        train_size=0.9,
        early_stopping=False,
        accelerator="gpu", devices=-1, strategy="ddp_find_unused_parameters_true",
        )

ori-kron-wis commented:
Hi @marianogabitto ,
Your code above should work in multiGPU settings, just add `distributed_sampler=True` to the DataSplitter call

Besides that I made several other updates for this PR, census/lamin custom dataloaders should be working now for scvi/scnavi/scarches/load/save/multiGPU/covariates integration
